a2s/a2s_async.py

```diff
@@ -1,138 +1,226 @@
+
+logger: logging.Logger = logging.getLogger("a2s")
+
+T = TypeVar("T", InfoProtocol, PlayersProtocol, RulesProtocol)
+
 
-logger = logging.getLogger("a2s")
+@overload
+async def request_async(
+    address: Tuple[str, int], timeout: float, encoding: str, a2s_proto: Type[InfoProtocol]
+) -> Union[SourceInfo, GoldSrcInfo]:
+    ...
 
 
-async def request_async(address, timeout, encoding, a2s_proto):
+@overload
+async def request_async(
+    address: Tuple[str, int], timeout: float, encoding: str, a2s_proto: Type[PlayersProtocol]
+) -> List[Player]:
+    ...
+
+
+@overload
+async def request_async(
+    address: Tuple[str, int], timeout: float, encoding: str, a2s_proto: Type[RulesProtocol]
+) -> Dict[Union[str, bytes], Union[str, bytes]]:
+    ...
+
+
+async def request_async(
+    address: Tuple[str, int], timeout: float, encoding: str, a2s_proto: Type[T]
+) -> Union[SourceInfo, GoldSrcInfo, List[Player], Dict[Union[str, bytes], Union[str, bytes]]]:
     conn = await A2SStreamAsync.create(address, timeout)
     response = await request_async_impl(conn, encoding, a2s_proto)
     conn.close()
     return response
 
-async def request_async_impl(conn, encoding, a2s_proto, challenge=0, retries=0, ping=None):
+
+@overload
+async def request_async_impl(
+    conn: A2SStreamAsync,
+    encoding: str,
+    a2s_proto: Type[InfoProtocol],
+    challenge: int = ...,
+    retries: int = ...,
+    ping: Optional[float] = ...,
+) -> Union[SourceInfo, GoldSrcInfo]:
+    ...
+
+
+@overload
+async def request_async_impl(
+    conn: A2SStreamAsync,
+    encoding: str,
+    a2s_proto: Type[PlayersProtocol],
+    challenge: int = ...,
+    retries: int = ...,
+    ping: Optional[float] = ...,
+) -> List[Player]:
+    ...
+
+
+@overload
+async def request_async_impl(
+    conn: A2SStreamAsync,
+    encoding: str,
+    a2s_proto: Type[RulesProtocol],
+    challenge: int = ...,
+    retries: int = ...,
+    ping: Optional[float] = ...,
+) -> Dict[Union[str, bytes], Union[str, bytes]]:
+    ...
+
+
+async def request_async_impl(
+    conn: A2SStreamAsync,
+    encoding: str,
+    a2s_proto: Type[T],
+    challenge: int = 0,
+    retries: int = 0,
+    ping: Optional[float] = None,
+) -> Union[SourceInfo, GoldSrcInfo, Dict[Union[str, bytes], Union[str, bytes]], List[Player]]:
     send_time = time.monotonic()
     resp_data = await conn.request(a2s_proto.serialize_request(challenge))
     recv_time = time.monotonic()
     # Only set ping on first packet received
     if retries == 0:
         ping = recv_time - send_time
 
-    reader = ByteReader(
-        io.BytesIO(resp_data), endian="<", encoding=encoding)
+    reader = ByteReader(io.BytesIO(resp_data), endian="<", encoding=encoding)
 
     response_type = reader.read_uint8()
     if response_type == A2S_CHALLENGE_RESPONSE:
         if retries >= DEFAULT_RETRIES:
-            raise BrokenMessageError(
-                "Server keeps sending challenge responses")
```

Comment:
Lines were originally broken here to keep their length below 79 characters, as recommended by PEP 8. Whether that limit is appropriate is up for discussion, but it certainly should not be changed in this PR.

Reply:
This was a stylistic preferece on my end - I think a line length of 79 is too limiting but it was rude of me to assume others would agree. This has been fixed (returned to 79 max line length) and will be pushed in a moment!

Reply:
Thank you!